code/main/media/MediaSession.brs

```diff
@@ -0,0 +1,266 @@
+            end if
+
+            ' Create and add hit to queue for actual events or heartbeat pings
+            mediaHit = m._createMediaHit(requestId, eventType, xdmData, tsObject)
+            m._hitQueue.append(mediaHit)
+            m._lastHit = mediaHit
+            m.processMediaEvents()
+        end sub,
+
+        processMediaEvents: sub()
+            if not m._hadValidConfig() then
+                _adb_logError("processMediaEvents() - Cannot process media event (" + FormatJson(eventType) + "), missing required configuration.")
+                return
+            end if
+
+            while _hitQueue.Count() <> 0
+                hit = _hitQueue.Shift()
+                requestId = hit.requestId
+                tsInMillis = hit.tsObject.tsInMillis
+
+                ' attach _id and timestamp in ISO format
+                xdmData.xdm["_id"] = _adb_generate_UUID()
+                xdmData.xdm["timestamp"] = hit.tsObject.tsInISO8601
+
+                ' attach sessionId to events other than sessionStart
+                if hit.eventType = m._CONSTANTS.MEDIA.EVENT_TYPE.SESSION_START
+                    xdmData = m._attachMediaConfig(xdmData)
+                    xdmData = m._updateChannelFromSessionConfig(xdmData)
+                else
+                    ' attach sessionId to events other than sessionStart
+                    if m._backendSessionId = invalid then
+                        _adb_logError("processQueuedEvents() - Cannot queue media event, backend session ID is not set.")
+                        return
+                    end if
+
+                    xdmData.xdm["mediaCollection"]["sessionID"] = m._backendSessionId
+                end if
+
+
+                xdm = [hit.xdmData]
+                path = m._MEDIA_PATH_PREFIX + hit.eventType
+                meta = {}
+
+                m._edgeRequestQueue.add(requestId, xdm, tsInMillis, meta, path)
+                m._processEdgeRequestQueue()
+            end while
+        end sub,
+
+        close: sub(isAbort as boolean = false)
+            _isActive = false
+
+            if isAbort then
+                ' Drop the hits in the queue
+                _hitQueue = []
+            else
+                ' Dispatch all the hits in the queue
+                m.processQueuedEvents()
+            end if
+        end sub,
+
+        getHitQueueSize: sub() as object
+            return _hitQueue.Count()
+        end sub,
+
+        handleError: sub(requestId as string, error as object)
+            ' Handle error
+            ' Drop the hits and mark session inactive if error with sessionStart
+        end sub,
+
+        _updateState: sub(requestId as string, eventType as string, xdmData as object, tsObject as object)
+            ' Update the session state based on the event type
+            m._updateAdState(eventType)
+            m._updatePlaybackState(eventType)
+            m._extractSessionStartData(eventType, xdmData)
+            m._closeIfIdle()
```

Comment:
We should also add some logic to restart when the playback state changes to playing from idle.